packages/xsnap/src/avaAssertXS.js

```diff
@@ -0,0 +1,315 @@
+/* eslint-disable no-await-in-loop */
+// @ts-check
+
+/** global print */
+
+const { freeze, keys } = Object;
+
+// ack Paul Roub Aug 2014
+// https://stackoverflow.com/a/25456134/7963
+/** @type {(x: unknown, y: unknown) => boolean } */
+function deepEqual(x, y) {
+  if (x === y) {
+    return true;
+  }
+  if (
+    typeof x === 'object' &&
+    x != null &&
+    typeof y === 'object' &&
+    y != null
+  ) {
+    if (keys(x).length !== keys(y).length) {
+      const detail = JSON.stringify({
+        actual: {
+          length: keys(x).length,
+          keys: keys(x),
+        },
+        expected: {
+          length: keys(y).length,
+          keys: keys(y),
+        },
+      });
+      throw new Error(`Object keys length: ${detail}`);
+    }
+
+    for (const prop in x) {
+      // eslint-disable-next-line no-prototype-builtins
+      if (y.hasOwnProperty(prop)) {
```

Comment:
If `y` happens to have a property names `'hasOwnPrototype'` this will invoke it, which is not what you meant. It looks like this is here as a vestige of ES3 era coding advice.